File: xiaoming/1004/ValidSudoku.java

public class ValidSudoku {
	
	public boolean isValidSudoku(char[][] board) {
	    if(board==null || board.length!=9 || board[0].length!=9)
	        return false;
	    for(int i=0;i<9;i++)
	    {
	        boolean[] map = new boolean[9];
	        for (int k = 0; k < 9; k++) map[k] = false;
	        for(int j=0;j<9;j++)
	        {
	            if(board[i][j] != '.')
	            {
	                if(map[(int)(board[i][j]-'1')])
	                {
	                    return false;
	                }
	                map[(int)(board[i][j]-'1')] = true;
	            }
	        }
	    }
	    
	    for(int j=0;j<9;j++)
	    {
	        boolean[] map = new boolean[9];
	        for (int k = 0; k < 9; k++) map[k] = false;
	        for(int i=0;i<9;i++)
	        {
	            if(board[i][j] != '.')
	            {
	                if(map[(int)(board[i][j]-'1')])
	                {
	                    return false;
	                }
	                map[(int)(board[i][j]-'1')] = true;
	            }
	        }
	    }        
	    for(int block=0;block<9;block++)
	    {
	        boolean[] map = new boolean[9];
	        for (int k = 0; k < 9; k++) map[k] = false;
	        for(int i=block/3*3;i<block/3*3+3;i++)
	        {
	            for(int j=block%3*3;j<block%3*3+3;j++)
	            {
	                if(board[i][j] != '.')
	                {
	                   if(map[(int)(board[i][j]-'1')])
	                   {
	                      return false;
	                   }
	                   map[(int)(board[i][j]-'1')] = true;    
	                }
	            }
	        }
	    }
	    return true;
	}

}

File: Sen/Valid_Sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        # validate rows
        for i in xrange(9):
            if board[i].count('.') + len(set(board[i])) - 1 != 9: # set(list) deletes all duplicates in the list and includes the element ".", that is why "- 1"
                return False
        # validate columns
        for i in xrange(9):
            col = [board[j][i] for j in xrange(9)]
            if col.count('.') + len(set(col)) - 1 != 9: 
                return False
        # validate 3x3 squares
        for i in (0, 3, 6): # for horizontal traversal
            for j in (0, 3, 6): # for vertical traversal  
                square = []
                for m in (0, 1, 2):
                    for n in (0, 1, 2):
                        square.append(board[i + m][j + n])
                # square = [board[i + m][j + n] for m in (0, 1, 2) for n in (0, 1, 2)]
                if square.count('.') + len(set(square)) - 1 != 9: 
                    return False
        return True 

board = [".87654321", "2........", "3........", "4........", "5........", "6........", "7........", "8........", "9........"]
i = Solution()
print i.isValidSudoku(board)

    